internal/internal_worker.go

```diff
@@ -457,22 +470,28 @@ func newActivityTaskWorker(taskHandler ActivityTaskHandler, service workflowserv
 	ensureRequiredParams(&workerParams)
 
 	poller := newActivityTaskPoller(taskHandler, service, workerParams)
-
+	workerType := "ActivityWorker"
+	metricsHandler := workerParams.MetricsHandler.WithTags(metrics.WorkerTags(workerType))
+	tss := newTrackingSlotSupplier(workerParams.Tuner.GetActivityTaskSlotSupplier(), metricsHandler)
 	base := newBaseWorker(
 		baseWorkerOptions{
 			pollerCount:       workerParams.MaxConcurrentActivityTaskQueuePollers,
 			pollerRate:        defaultPollerRate,
-			maxConcurrentTask: workerParams.ConcurrentActivityExecutionSize,
+			slotSupplier:      tss,
 			maxTaskPerSecond:  workerParams.WorkerActivitiesPerSecond,
 			taskWorker:        poller,
 			identity:          workerParams.Identity,
-			workerType:        "ActivityWorker",
+			workerType:        workerType,
 			stopTimeout:       workerParams.WorkerStopTimeout,
 			fatalErrCb:        workerParams.WorkerFatalErrorCallback,
 			userContextCancel: workerParams.UserContextCancel,
+			metricsHandler:    metricsHandler,
+			slotReservationData: slotReservationData{
+				taskQueue: workerParams.TaskQueue,
+				logger:    workerParams.Logger,
+			},
 		},
 		workerParams.Logger,
-		workerParams.MetricsHandler,
```

Comment:
Why move just the `MetricsHandler` into the options? TBH I am not sure why `newBaseWorker` takes an options struct and parameters?

Reply:
I dunno, why it takes both - just already was that way.. I moved the metric handler in because it needs to be in the base worker options because it needed to be tagged before calling `newBaseWorker` now, and in the process I observed what you did and figured there was no reason for it to be a separate option.

I could also just move the other parameters in too if we want. I only didn't since I didn't touch them.

Reply:
Yeah it's just now a bit silly to pass `workerType` in at all because your tagging the metric handler outside `newBaseWorker`, but tagging the logger with the same tag in `newBaseWorker`

Reply:
Why does it need to be tagged before calling `newBaseWorker`?  Why can't we just construct `newTrackingSlotSupplier ` inside `newBaseWorker`?

Reply:
It makes it easier for tests that call `newBaseWorker` directly to use a specific slot supplier.

But, yeah I can just get rid of the other params to newBaseWorker and move workerType out of the options.

Reply:
We are repeating the same basic code 4 times for each worker, i would prefer to optimize for the actual SDK code and not tests

Reply:
In the SDK `newBaseWorker` is only used `newTrackingSlotSupplier` so i think this common code should just be in `newBaseWorker`

Reply:
Sure, changed.

Reply:
It still looks like we are constructing the logger and metric handler outside `newBaseWorker`

Reply:
Sorry, I misread you. I read it as you not wanting to pass _in_ workerType, hence them needing to be outside, but yeah I agree it'd be better the other way around.